# DX 602 Final Project

## Introduction

In this project, you will practice the skills that you have learned throughout this module with a heavy focus on building models.
Most of the problems and questions are open ended compared to your previous homeworks, and you will be asked to explain your choices.
Most of them will have a particular type of solution implied, but it is up to you to figure out the details based on what you have learned in this module.

## Instructions

Each problem asks you to perform build models, run a computation, or otherwise perform some analysis of the data, and usually answer some questions about the results.
Make sure that your question answers are well supported by your analysis and explanations; simply stating an answer without support will earn minimal points.

Notebook cells for code and text have been added for your convenience, but feel free to add additional cells.

## Example Code

You may find it helpful to refer to this GitHub repository of Jupyter notebooks for example code.

* https://github.com/bu-cds-omds/dx601-examples
* https://github.com/bu-cds-omds/dx602-examples

Any calculations demonstrated in code examples or videos may be found in these notebooks, and you are allowed to copy this example code in your homework answers.

## Submission

This project will be entirely manually graded.
However, we may rerun some or all of your code to confirm that it works as described.

### Late Policy

The normal homework late policy for OMDS does not apply to this project.
Boston University requires final grades to be submitted within 72 hours of class instruction ending, so we cannot accommodate 5 days of late submissions.

However, we have delayed the due date of this project to be substantially later than necessary given its scope, and given you more days for submission with full credit than you would have had days for submission with partial credit under the homework late policy.
The Thanksgiving holiday was also taken into account in setting the deadline.
Finally, the deadlines for DX 601 and DX 602 were coordinated to be a week apart while giving ample time for both of their projects.

## Shared Imports

For this project, you are forbidden to use modules that were not loaded in this template.
While other modules are handy in practice, modules that trivialize these problems interfere with our assessment of your own knowledge and skills.

If you believe a module covered in the course material (not live sessions) is missing, please check with your learning facilitator.

In [1]:
import math
import sys

In [2]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import scipy.stats
import sklearn.linear_model

## Problems

### Problem 1 (5 points)

Pick one of the following data sets to analyze in this project.
Load the data set, and show a random sample of 10 rows.

* [Wine Quality](https://archive.ics.uci.edu/dataset/186/wine+quality) ([PMLB - red subset only](https://github.com/EpistasisLab/pmlb/tree/master/datasets/wine_quality_red))
* [Body Fat](https://www.openml.org/search?type=data&status=active&id=560) ([PMLB](https://github.com/EpistasisLab/pmlb/tree/master/datasets/560_bodyfat))

The PMLB copies of the data are generally cleaner and recommended for this project, but the other links are provided to give you more context.
To load the data from the PMLB Github repository, navigate to the `.tsv.gz` file in GitHub and copy the link from the "Raw" button.

If the dataset has missing data, you should drop the rows with missing data before proceeding.
If the data set you choose has more than ten columns, you may limit later analysis that is requested per column to just the first ten columns.

In [12]:
# YOUR CODE HERE

# Loading the dataset
bodyFat_dataset_url = "https://github.com/EpistasisLab/pmlb/raw/refs/heads/master/datasets/560_bodyfat/560_bodyfat.tsv.gz"
df_bodyFat = pd.read_csv(bodyFat_dataset_url, sep="\t")

# Count missing values
null_per_column = df_bodyFat.isnull().sum()
total_null = df_bodyFat.isnull().sum().sum()
print("----------------------------------------------")
print(f"Count of missing values per column : \n{null_per_column}")
print("----------------------------------------------")
print(f"Total number of missing values in the dataset = {total_null}")

# Drop rows with missing values if they exist 
df_bodyFat = df_bodyFat.dropna()

# Dataset shape
print("----------------------------------------------")
print(f"Shape of the dataset:")
print(f"> Number of rows = {df_bodyFat.shape[0]}")
print(f"> Number of columns = {df_bodyFat.shape[1]}")
print("----------------------------------------------")

# Show a random sample of 10 rows
df_bodyFat_sample = df_bodyFat.sample(10, random_state=2025)
print("---------- random sample of 10 rows ----------\n")
df_bodyFat_sample


----------------------------------------------
Count of missing values per column : 
Density    0
Age        0
Weight     0
Height     0
Neck       0
Chest      0
Abdomen    0
Hip        0
Thigh      0
Knee       0
Ankle      0
Biceps     0
Forearm    0
Wrist      0
target     0
dtype: int64
----------------------------------------------
Total number of missing values in the dataset = 0
----------------------------------------------
Shape of the dataset:
> Number of rows = 252
> Number of columns = 15
----------------------------------------------
---------- random sample of 10 rows ----------



Density   Age  Weight  Height       Neck       Chest    Abdomen  \
14    1.0484  35.0  187.75   69.50  40.500000  101.300003  96.400002   
187   1.0524  41.0  210.50   72.00  38.500000  107.400002  98.900002   
83    1.0377  70.0  170.75   70.00  38.700001  101.800003  94.900002   
48    1.0678  45.0  135.75   68.50  32.799999   92.300003  83.400002   
56    1.0524  58.0  181.50   68.00  39.099998  100.000000  99.800003   
80    1.0283  67.0  163.75   67.75  38.400002   97.699997  95.800003   
133   1.0398  50.0  157.00   66.75  37.799999  100.400002  89.400002   
30    1.0716  32.0  182.00   73.75  38.700001  100.500000  88.699997   
200   1.0709  43.0  178.25   70.25  37.799999  102.699997  89.199997   
161   1.0653  33.0  196.00   73.00  38.500000  103.800003  95.599998   

            Hip      Thigh       Knee      Ankle     Biceps    Forearm  \
14   100.099998  69.000000  39.000000  23.100000  36.099998  30.500000   
187  104.099998  63.500000  39.799999  23.500000  36.400002  30.400000   
83    95.000000  56.000000  36.500000  24.100000  31.200001  27.299999   
48    90.400002  52.000000  35.799999  20.600000  28.799999  25.500000   
56   102.500000  62.099998  39.599998  22.500000  33.099998  28.299999   
80    97.099998  54.799999  38.200001  23.700001  29.400000  27.200001   
133   92.300003  56.099998  35.599998  20.500000  33.599998  29.299999   
30    99.800003  57.500000  38.700001  33.900002  32.500000  27.700001   
200   99.199997  60.200001  39.200001  23.799999  31.700001  28.400000   
161  105.099998  61.400002  40.599998  25.000000  31.299999  29.200001   

         Wrist  target  
14   18.200001    22.1  
187  19.100000    20.4  
83   19.200001    27.0  
48   16.299999    13.6  
56   18.500000    20.4  
80   19.000000    31.4  
133  17.299999    26.1  
30   18.400000    11.9  
200  18.600000    12.2  
161  19.100000    14.6

For this final project, I selected the **Body Fat** dataset. This dataset originates from a study on estimating <u>body fat</u> percentage using various <u>body measurements</u>.<br>
The purpose is to explore whether inexpensive, easy-to-obtain measurements can reliably predict a person's body fat. <br><br>
The **Body Fat** dataset contains:
* Number of rows: 252
* Number of columns: 15 (All columns in the dataset are <u>numeric</u>) 


**Missing Values Analysis** <br>
The dataset was inspected for missing values:
* Missing values per column: all zeros
* Total missing values: 0

=> Since no missing values were found, no rows needed to be removed.<br>

**The sample of 10 rows** <br>
After loading the dataset and displaying a random sample of 10 rows, I can observe that the dataset includes 15 columns:
* 14 numerical feature columns (Predictors): measurements such as *Height* and *Hip*
* 1 target column that represents the body fat percentage (continuous numerical variable)

### Problem 2 (10 points)

List all of the columns and describe them in your own words.

In [ ]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 3 (50 points)

Perform an exploratory analysis of the data set.
After your exploratory analysis, pick 3 individual charts that you the think were particularly interesting.
Repeat those charts separately from your original analysis, and after each of those charts, explain what you thought was noteworthy.

In [ ]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 4 (5 points)

Plot the correlation matrix of the numeric columns in the data set.
Which pair of different columns were highlighted as the most correlated?

In [ ]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 5 (10 points)

Pick three different regression model classes to try in problem 6 from the scikit-learn documentation.
For each class, provide a link to the scikit-learn documentation, and a link to another web page describing how that kind of model works.
The second link should not be from scikit-learn, but Wikipedia is acceptable.
You do not need to understand the methods at this time, but it is good to be comfortable researching them.

In [ ]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 6 (50 points)

Build three different regression models using the entire data set.
Plot the actual target vs the predicted values for each in one chart.
Compute the L2 and L1 losses for each of them.
You may use any regression class provided provided by scikit-learn, and you may reuse one class as long as you change its parameters enough to see different results.

In [ ]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 7 (30 points)

Use 5-fold cross-validation to repeat building the same three kinds of regression models. Compare the L2 losses predicted by cross-validation against the L2 losses training against the whole data set. (The difference is likely from overfitting in the latter.)

In [ ]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 8 (25 points)

Build three different regression models as in problem 6, but preprocess the data so that each column has mean zero and standard deviation one first.
For full credit, use a scikit-learn pipeline for each model.
For each model, compare the L2 losses -- which of them performed differently from your results in problem 6?

(This process will be covered in week 13.)

In [ ]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 9 (5 points)

A colleague suggests that you find better models by repeatedly building decision trees with random depth limits.
They say that trying 1000 such models will likely find an improvement as long as you use cross validation.
Give a one sentence response to this suggestion. 

In [ ]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 10 (10 points)

Pick a best model from all the models that you built and otherwise described in this project.
Explain how you picked it, including what criteria you chose, and how the other models compared by that criteria.
As much as possible, justify that problem in the context of the original data set. 

In [ ]:
# YOUR CODE HERE

YOUR ANSWERS HERE